src/AspNetCore/Src/inprocessapplication.cxx

```diff
@@ -269,30 +408,141 @@ ASPNETCORE_APPLICATION::Initialize(
     // It all timed out
     if (dwResult == WAIT_TIMEOUT)
     {
-        return HRESULT_FROM_WIN32(dwResult);
+        // do we need kill the backend thread
+        hr = HRESULT_FROM_WIN32(dwResult);
+        goto Finished;
     }
     else if (dwResult == WAIT_FAILED)
     {
-        return HRESULT_FROM_WIN32(GetLastError());
+        hr = HRESULT_FROM_WIN32(GetLastError());
+        goto Finished;
     }
 
-    dwResult = WaitForSingleObject(m_hThread, 0);
-
     // The thread ended it means that something failed
     if (dwResult == WAIT_OBJECT_0)
     {
-        return HRESULT_FROM_WIN32(dwResult);
+        hr = E_APPLICATION_ACTIVATION_EXEC_FAILURE;
+        goto Finished;
     }
-    else if (dwResult == WAIT_FAILED)
+
+    m_fManagedAppLoaded = TRUE;
+
+Finished:
+    if (fLocked)
+    {
+        ReleaseSRWLockExclusive(&m_srwLock);
+    }
+
+    if (FAILED(hr))
+    {
+        // Question: in case of application loading failure, should we allow retry on 
+        // following request or block the activation at all
+        m_fLoadManagedAppError = FALSE; // m_hThread != NULL ?
+
+        if (SUCCEEDED(strEventMsg.SafeSnwprintf(
+            ASPNETCORE_EVENT_LOAD_CLR_FALIURE_MSG,
+            m_pConfiguration->QueryApplicationPath()->QueryStr(),
+            m_pConfiguration->QueryApplicationFullPath()->QueryStr(),
+            hr)))
+        {
+            apsz[0] = strEventMsg.QueryStr();
+
+            //
+            // not checking return code because if ReportEvent
+            // fails, we cannot do anything.
+            //
+            if (FORWARDING_HANDLER::QueryEventLog() != NULL)
+            {
+                ReportEventW(FORWARDING_HANDLER::QueryEventLog(),
+                    EVENTLOG_ERROR_TYPE,
+                    0,
+                    ASPNETCORE_EVENT_LOAD_CLR_FALIURE,
+                    NULL,
+                    1,
+                    0,
+                    apsz,
+                    NULL);
+            }
+        }
+    }
+    return hr;
+}
+
+VOID
+IN_PROCESS_APPLICATION::Recycle(
+    VOID
+)
+{
+    DWORD    dwThreadStatus = 0;
+    DWORD    dwTimeout = m_pConfiguration->QueryShutdownTimeLimitInMS();
+
+    AcquireSRWLockExclusive(&m_srwLock);
+    if (!g_fRecycleProcessCalled)
+    {
+        g_fRecycleProcessCalled = TRUE;
+
+        // notify IIS first so that new request will be routed to new worker process
+        g_pHttpServer->RecycleProcess(L"AspNetCore Recycle Process on Demand");
+    }
+    // First call into the managed server and shutdown
+    if (m_ShutdownHandler != NULL)
+    {
+        m_ShutdownHandler(m_ShutdownHandlerContext);
+        m_ShutdownHandler = NULL;
+    }
+
+    if (m_hThread != NULL &&
+        GetExitCodeThread(m_hThread, &dwThreadStatus) != 0 &&
+        dwThreadStatus == STILL_ACTIVE)
+    {
+        // wait for gracefullshut down, i.e., the exit of the background thread or timeout
+        if (WaitForSingleObject(m_hThread, dwTimeout) != WAIT_OBJECT_0)
```

Comment:
Have we confirmed this succeeds with an application today? Does terminate thread get called (hopefully not)?